harness/determined/deploy/gke/cli.py

```diff
@@ -0,0 +1,596 @@
+    if not pod_spec:
+        pod_spec = {"apiVersion": "v1", "kind": "Pod"}
+    pod_spec = cast(Dict[str, Union[Dict, str]], pod_spec)
+    if not pod_spec.get("spec"):
+        pod_spec["spec"] = {}
+    return pod_spec
+
+
+def validate_accelerator_type(s: str) -> None:
+    json_value = subprocess.check_output(
+        ["gcloud", "compute", "accelerator-types", "list", "--format=json(name)"]
+    )
+    json_names = json.loads(json_value)
+    valid_accelerator_types = {accelerator["name"] for accelerator in json_names}
+
+    if s not in valid_accelerator_types:
+        raise ValueError("Accelerator must be one of {}".format(valid_accelerator_types))
+
+
+def validate_location(location: str, isZone: bool = True) -> None:
+    try:
+        cmd = ["gcloud", "compute"]
+        if isZone:
+            cmd += ["zones"]
+        else:
+            cmd += ["regions"]
+        cmd += ["describe", location]
+        subprocess.check_call(cmd, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
+    except subprocess.CalledProcessError:
+        raise ValueError(
+            "The specified {} {} was not found".format("zone" if isZone else "region", location)
+        )
+
+
+def validate_machine_type(machine_type: str, zone: str) -> None:
+    try:
+        subprocess.check_call(
+            [
+                "gcloud",
+                "compute",
+                "machine-types",
+                "describe",
+                machine_type,
+                "--zone",
+                zone,
+            ],
+            stdout=subprocess.DEVNULL,
+        )
+    except subprocess.CalledProcessError:
+        raise ValueError(
+            "The specified machine type {} was not found in the specified zone {}.".format(
+                machine_type, zone
+            ),
+        )
+
+
+def validate_accelerator_zone(args: argparse.Namespace, zone: str) -> None:
+    try:
+        subprocess.check_call(
+            [
+                "gcloud",
+                "compute",
+                "accelerator-types",
+                "describe",
+                args.gpu_type,
+                "--zone",
+                zone,
+            ],
+            stdout=subprocess.DEVNULL,
+        )
+    except subprocess.CalledProcessError:
+        raise ValueError(
+            "The specified accelerator type {} was not found in the specified zone {}.".format(
+                args.gpu_type, zone
+            ),
+        )
+
+
+def validate_args(args: argparse.Namespace) -> None:
+    validate_location(args.zone, isZone=True)
+    validate_accelerator_type(args.gpu_type)
+
+    if args.master_machine_type != "n1-standard-16":
+        validate_machine_type(args.master_machine_type, args.zone)
+
+    if args.agent_machine_type != "n1-standard-32":
+        validate_machine_type(args.agent_machine_type, args.zone)
+
+    validate_accelerator_zone(args, args.zone)
+
+    if args.gpu_coscheduler and args.preemption:
```

Comment:
This predates me but my understanding is that the goal was to allow configuring coscheduler on GPU pool only while having two pools.

Technically we have distributed CPU jobs, although I doubt there're any users on k8s using it.

Reply:
Ah - yeah I always interpret the "CPU training" work as meaning something other than what it means :smile:

Reply:
The split is separate flags to enable the scheduler for the GPU / CPU resource pools.

Reply:
hmm well it makes sense to check for whether the coscheduler and the preemption scheduler are both set (since only one should be set at a time), but it is not currently possible to run a CPU pool alongside a GPU pool.

Reply:
I believe the case which is handled here is that you may have slotted jobs running on GPUs with coscheduler; and then there's CPU-only nodes without coscheduler for zero-slot tasks; and there's some amount of code to setup taints/tolerations here to ensure that stuff is getting scheduled that way.

If you also want coscheduler on CPU's, this code will also force you to have coscheduler on GPUs as well. So that case is covered.

I could rip it out if you think it doesn't make sense. 

But in this tool I'd try to lean towards being more flexible than having guardrails and only allowing safe and sound setups. This thing is not recommended for production really. I see it as a development tool. So that if you're testing something, you could setup a configuration using some combination of flags, and give that combination of flags to other people - which is better than having to do all this setup manually (where you've got to run 5+ pretty long command lines).

Reply:
alright, I've ripped it out and rebased the whole branch. PTAL.